File: services/Services.BackupCreator/Jobs/ZipBackupCreator.cs
using Microsoft.Extensions.Logging;
using Services.BackupCreator.Config;
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Text;

namespace Services.BackupCreator.Jobs
{
    public class ZipBackupCreator
    {
        private readonly ILogger _logger;
        private readonly PathsConfig _pathsConfig;

        public ZipBackupCreator(ILogger<BackupCreatorService> logger, PathsConfig pathsConfig)
        {
            _logger = logger;
            _pathsConfig = pathsConfig;
        }

        public string CreateBackup()
        {
            _logger.LogInformation("Starting to create zip");

            var backupDir = GetBackupDir();
            CopyDirsToBackup(backupDir);
            string zipFilePath = CreateZipFilePath(backupDir);
            CreateZipFile(backupDir, zipFilePath);

            return zipFilePath;
        }

        private void CreateZipFile(DirectoryInfo backupDir, string zipFilePath)
        {
            ZipFile.CreateFromDirectory(Path.Combine(backupDir.FullName, "create"), zipFilePath,
                            CompressionLevel.Optimal, false);

            var fileSize = new System.IO.FileInfo(zipFilePath).Length;
            var fileSizeKB = fileSize / 1024 * 1.0;
            var fileSizeMB = Math.Truncate(fileSizeKB / 1024 * 1000) / 1000;

            _logger.LogInformation("Created zip: {zipFilePath}. File size: {fileSizeMB}", zipFilePath, fileSizeMB);
        }

        private string CreateZipFilePath(DirectoryInfo backupDir)
        {
            var now = DateTime.Now;
            var nowString = now.ToString("yyyy-MM-dd-HHmmss");
            var backupFileName = $"backup_" + nowString + ".zip";

            var zipFilePath = Path.Combine(backupDir.FullName, backupFileName);
            if (File.Exists(zipFilePath))
                File.Delete(zipFilePath);
            return zipFilePath;
        }

        private void CopyDirsToBackup(DirectoryInfo backupDir)
        {
            foreach (var dir in _pathsConfig.DirsToBackup)
            {
                var sourceDir = Path.GetFullPath(dir);

                if (!Directory.Exists(sourceDir))
                {
                    _logger.LogWarning("Dir {sourceDir} does not exist", sourceDir);
                    continue;
                }

                var dirName = Path.GetFileName(sourceDir);

                var destinationDir = Path.Combine(backupDir.FullName, "create", dirName);
                Directory.CreateDirectory(destinationDir);

                PerformDeepCopy(sourceDir, destinationDir);

                _logger.LogInformation("Copied {sourceDir} to {destinationDir}", sourceDir, destinationDir);
            }
        }

        private DirectoryInfo GetBackupDir()
        {
            var path = Path.GetFullPath(_pathsConfig.TempBackupPath);
            var backupDir = Directory.CreateDirectory(path);
            backupDir.Delete(true);

            return backupDir;
        }

        private void PerformDeepCopy(string sourceDirectory, string destinationDirectory)
        {
            if (!Directory.Exists(destinationDirectory))
            {
                Directory.CreateDirectory(destinationDirectory);
            }
            var sourceDir = new DirectoryInfo(sourceDirectory);
            var targetDir = new DirectoryInfo(destinationDirectory);
            Array.ForEach(sourceDir.GetFiles(), f => f.CopyTo(Path.Combine(targetDir.ToString(), f.Name), true));

            foreach (DirectoryInfo dir in sourceDir.GetDirectories())
            {
                var subDirectory = targetDir.CreateSubdirectory(dir.Name);
                PerformDeepCopy(dir.FullName, subDirectory.FullName);
            }
        }
    }
}

File: services/SiteParser/create_backup_job/zip_backup_creator.py
import os
from os import path
from distutils.dir_util import copy_tree
import datetime
import zipfile
from configs.paths_config import PathsConfig
import shutil


class ZipBackupCreator:
    paths_config = PathsConfig()

    def create_backup(self):
        print("Starting to create zip")

        backup_dir = self.__get_backup_dir()
        self.__copy_dirs_to_backup(backup_dir)
        zip_file_path = self.__get_zip_file_path(backup_dir)
        self.__create_zip_file(backup_dir, zip_file_path)

        return zip_file_path

    @staticmethod
    def __get_full_path(pth):
        return path.normpath(path.join(os.getcwd(), pth))

    def __get_backup_dir(self):
        backup_dir = self.__get_full_path(self.paths_config.temp_backup_path)
        if path.exists(backup_dir):
            shutil.rmtree(backup_dir)

        return backup_dir

    def __copy_dirs_to_backup(self, backup_dir):
        for directory in self.paths_config.dirs_to_backup:
            source_dir = self.__get_full_path(directory)
            if not path.exists(source_dir):
                print("Dir {0} does not exist".format(source_dir))

            dir_name = path.basename(source_dir)
            destination_dir = path.join(backup_dir, "create", dir_name)

            self.__perform_deep_copy(source_dir, destination_dir)

            print("Copied {source} to {dest}".format(source=source_dir, dest=destination_dir))

    @staticmethod
    def __create_dir(pth):
        os.makedirs(pth, exist_ok=True)

    def __perform_deep_copy(self, source_dir, destination_dir):
        self.__create_dir(destination_dir)
        copy_tree(source_dir, destination_dir)

    @staticmethod
    def __get_zip_file_path(backup_dir):
        date_string = datetime.datetime.now().strftime("%Y-%m-%d-%H%M%S")
        backup_file_name = "Win_backup_{0}.zip".format(date_string)
        zip_file_path = path.join(backup_dir, backup_file_name)
        if os.path.exists(zip_file_path):
            os.remove(zip_file_path)

        return zip_file_path

    @staticmethod
    def __create_zip_file(backup_dir, zip_file_path):
        with zipfile.ZipFile(zip_file_path, 'w', zipfile.ZIP_DEFLATED) as zip:
            for root, dirs, files in os.walk(path.join(backup_dir, "create")):
                for file in files:
                    zip.write(path.join(root, file))

        file_size_b = path.getsize(zip_file_path)
        file_size_kb = file_size_b / 1024
        file_size_mb = file_size_kb / 1024
        print("Created zip: {zipFilePath}. File size: {fileSizeMB:.3f}"
              .format(zipFilePath=zip_file_path, fileSizeMB=round(file_size_mb, 3)))
